### PR TITLE
fix: providing separate agents for HTTP and HTTPS proxy servers

### PR DESCRIPTION
**Pull request by rupareddy5-21**
creating separate http and https agents depending on whether:
The user is hitting an http:// or https:// dev server.
The user is starting the CLI using --ssl or --no-ssl (by default)
Also, warning the user if:
They started the CLI using  --ssl but they are connecting to http:// dev server
They started the CLI with no SSL but they are connecting to https:// dev server

### CODE REVIEW

File: src/cli/commands/start.ts [hunk 1]
```diff
@@ -112,6 +116,7 @@ swa start http://localhost:3000 --api-devserver-url http://localhost:7071
     );
 }
 
+export var isSSL: boolean = false;
```
Comment (manekinekko) [inline src/cli/commands/start.ts:119]: commands should not export variables. Use `swaCLIEnv()` for that.

File: src/msha/server.ts [hunk 2]
```diff
@@ -15,16 +16,30 @@ import { handleUserConfig, onConnectionLost, requestMiddleware } from "./middlew
...
+} else {
+  proxyApp = httpProxy.createProxyServer({
+    autoRewrite: true,
+    agent: new http.Agent({
+      keepAlive: true,
+      keepAliveMsecs: 5000,
+    }),
+  });
+  if (isHttpsUrl(SWA_CLI_API_URI())) {
+    logger.error(`Your connection is not secure. Please start the CLI using the flag --ssl. Exiting`, true);
```
Comment (manekinekko) [inline src/msha/server.ts:41]: We are introducing a breaking change. Please add this new behavior to the docs.

File: src/msha/server.ts [hunk 3]
```diff
@@ -40,6 +55,19 @@ const httpsServerOptions: Pick<https.ServerOptions, "cert" | "key"> | null =
...
+  }
+
+  try {
+    const uri = new URL(url);
+    return uri.protocol.startsWith("https");
+  } catch {
+    return false;
+  }
+}
+
```
Comment (manekinekko) [inline src/msha/server.ts:70]: move this helper function under `core/...`

**Review by manekinekko — CHANGES_REQUESTED** [review]

File: src/msha/server.ts [hunk 4]
```diff
@@ -40,6 +55,19 @@ const httpsServerOptions: Pick<https.ServerOptions, "cert" | "key"> | null =
       }
     : null;
 
+function isHttpsUrl(url: string | undefined) {
```
Comment (sulabh-msft) [inline src/msha/server.ts:58]: Also can we add test for this method if it moving to core.

File: src/core/utils/net.spec.ts [hunk 5]
```diff
@@ -310,4 +310,13 @@ describe("net utilities", () => {
...
     });
   });
+
+  describe("isHttpsUrl()", () => {
+    it("https url should be valid", () => {
+      expect(isHttpsUrl("https://foo.com")).toBeTrue();
+    });
+    it("http url should be invalid", () => {
+      expect(isHttpsUrl("http://foo.com")).toBeFalse();
+    });
```
Comment (manekinekko) [inline src/core/utils/net.spec.ts:320]: please also add invalid values to make sure the code won't throw unexpected errors.

**Review by manekinekko — CHANGES_REQUESTED** [review]

File: docs/www/docs/cli/swa-start.md [hunk 6]
```diff
@@ -160,6 +160,10 @@ Connect both front-end and the API to running development server
 swa start http://localhost:3000 --api-devserver-url http://localhost:7071
 ```
 
+:::info Note
```
Comment (sulabh-msft) [inline docs/www/docs/cli/swa-start.md:163]: just for my understanding how will this section be rendered ?

File: src/core/utils/net.spec.ts (hunk 5 above)
Comment (sulabh-msft) [reply src/core/utils/net.spec.ts:320]: +1
Scenarios:
1. Null url
2. invalid url ( e..g random string  .. xysaa)
3. url with caps protocol - HTTPS4.

**Review by sulabh-msft — APPROVED** [review]

**Comment by rupareddy5-21** [comment]
Fixes #610

**Comment by manekinekko** [comment]
Great job @rupareddy5-21 🎉

**Pull request merged**